src/identity/hcs/did/hcs-did.ts

```diff
@@ -0,0 +1,337 @@
+            this.idString = HcsDid.publicKeyToIdString(didRootKey);
+            this.did = this.buildDid();
+
+            return;
+        }
+
+        if (
+            (typeof args[0] === 'string') &&
+            (args[1] instanceof PrivateKey) &&
+            (args[2] instanceof FileId) &&
+            ((args[4] instanceof TopicId) || args[4] === undefined) &&
+            (args.length === 4)
+        ) {
+            const [network, privateDidRootKey, addressBookFileId, didTopicId] = args
+
+            this.didTopicId = didTopicId;
+            this.addressBookFileId = addressBookFileId;
+            this.network = network;
+            this.didRootKey = privateDidRootKey.publicKey;
+            this.idString = HcsDid.publicKeyToIdString(privateDidRootKey.publicKey);
+            this.did = this.buildDid();
+            this.privateDidRootKey = privateDidRootKey;
+
+            return;
+        }
+
+        if (
+            (typeof args[0] === 'string') &&
+            (args[1] instanceof PublicKey) &&
+            (args[2] instanceof FileId) &&
+            (args.length === 3)
+        ) {
+            const [network, didRootKey, addressBookFileId] = args;
+
+            this.didTopicId = null;
+            this.addressBookFileId = addressBookFileId;
+            this.network = network;
+            this.didRootKey = didRootKey;
+            this.idString = HcsDid.publicKeyToIdString(didRootKey);
+            this.did = this.buildDid();
+
+            return;
+        }
+
+        if (
+            (typeof args[0] === 'string') &&
+            (typeof args[1] === 'string') &&
+            (args[2] instanceof FileId) &&
+            ((args[4] instanceof TopicId) || args[4] === undefined) &&
+            (args.length === 4)
+        ) {
+            const [network, idString, addressBookFileId, didTopicId] = args;
+
+            this.didTopicId = didTopicId;
+            this.addressBookFileId = addressBookFileId;
+            this.network = network;
+
+            this.idString = idString;
+            this.did = this.buildDid();
+
+            return;
+        }
+
+        throw new Error('Couldn\'t find constructor');
+    }
+
+    /**
+     * Converts a Hedera DID string into {@link HcsDid} object.
+     *
+     * @param didString A Hedera DID string.
+     * @return {@link HcsDid} object derived from the given Hedera DID string.
+     */
+    public static fromString(didString: string): HcsDid {
+        if (!didString) {
+            throw new Error("DID string cannot be null");
+        }
+
+        const mainParts = didString.split(DidSyntax.DID_PARAMETER_SEPARATOR);
+
+        let
+            topicId: TopicId,
+            addressBookFileId: FileId;
+
+        try {
+            const didParts = mainParts.shift().split(DidSyntax.DID_METHOD_SEPARATOR);
+
+            if (didParts.shift() !== DidSyntax.DID_PREFIX) {
+                throw new Error('DID string is invalid: invalid prefix.');
```

Comment:
Using exceptions for local control flow is usually discouraged. I think you have it structured this way because you have a few specific errors that you want to throw, but you also want to catch any others that may happen in other calls, wrap those, and rethrow them. I'm not sure of a better solution.

Reply:
Actually we did this for modify error message for all messages. JS cannot catch specific errors like c++ or java. so we can only catch all, look into error and decide if we want rethrow it. Anyway i did some changes and move this errors outside of the block.